### PR TITLE
Update the correct abd in l2arc_read_done()

### PR DESCRIPTION
**Pull request by tcaputi**
This patch fixes an issue where l2arc_read_done() would always
write data to b_pabd, even if raw encrypted data was requested.
This only occured in cases where the L2ARC device had a different
ashift than the main pool.

Fixes: #7586

Signed-off-by: Tom Caputi <tcaputi@datto.com>

### How Has This Been Tested?
I will try to test this soon. It's a bit tricky since it requires having a L2ARc device with a different ashift from the rest of the pool. I'll probably hack some temporary test into the code, but I'm not sure about a long-term solution.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

**Comment by tcaputi** [comment]
@ahrens Can you see if this patch fixes the problem for you and Gernot? I will try to hack together a test on Monday.

File: module/zfs/arc.c [hunk 1]
```diff
@@ -8178,8 +8178,14 @@ l2arc_read_done(zio_t *zio)
 	if (cb->l2rcb_abd != NULL) {
 		ASSERT3U(arc_hdr_size(hdr), <, zio->io_size);
 		if (zio->io_error == 0) {
-			abd_copy(hdr->b_l1hdr.b_pabd, cb->l2rcb_abd,
-			    arc_hdr_size(hdr));
+			if (BP_IS_ENCRYPTED(&cb->l2rcb_bp) &&
+			    (cb->l2rcb_flags & ZIO_FLAG_RAW_ENCRYPT)) {
```
Comment (ahrens) [inline module/zfs/arc.c]: This change looks good to me.

I think there's also an opportunity to consolidate how we determine whether we're using b_rabd vs b_pabd.  We figure that out in 3 different places (here, with the same code on line 8205, and then when we calculate using_rdata).  Maybe we should do something like this early in the func, and then just use it:
```
boolean_t using_rdata =
    BP_IS_ENCRYPTED(&cb->l2rcb_bp) && (cb->l2rcb_flags & ZIO_FLAG_RAW_ENCRYPT);
```

Comment (tcaputi) [reply module/zfs/arc.c]: Fair enough. Fixed.

**Comment by tcaputi** [comment]
@ahrens We ran a manual test that consisted of:

1. Alter the L2ARC code so that it believes L2ARC devices always have an ashift of 13.
2. Insert the module with 32M min arc size and 64M max arc size
3. Create a pool with an ashift of 12 and an L2ARC device.
4. Clone and compile the ZFS codebase inside the root dataset of the pool
5. Snapshot and perform a raw send of the dataset (confirming this incurrned L2ARC hits.

Without this patch, the system crashed. With it, the send worked normally. I'm not sure how we can implement this into the automated test suite, but I think this at least confirms that the change is working as intended.

**Pull request merged**